src/Projection/InMemoryProjectionManager.php

```diff
@@ -112,48 +112,63 @@ public function stopProjection(string $name): void
         throw new Exception\RuntimeException('Stopping a projection is not supported in ' . get_class($this));
     }
 
-    public function fetchProjectionNames(?string $filter, bool $regex, int $limit, int $offset): array
+    public function fetchProjectionNames(?string $filter, int $limit = 20, int $offset = 0): array
     {
-        if (null === $filter && $regex) {
-            throw new Exception\InvalidArgumentException('No regex pattern given');
+        if (0 >= $limit) {
+            throw new Exception\OutOfRangeException(
+                'Invalid limit "'.$limit.'" given. Must be greater than 0.'
+            );
         }
 
-        if ($regex && false === @preg_match("/$filter/", '')) {
-            throw new Exception\InvalidArgumentException('Invalid regex pattern given');
+        if (0 > $offset) {
+            throw new Exception\OutOfRangeException(
+                'Invalid offset "'.$offset.'" given. Must be greater or equal than 0.'
+            );
         }
 
-        $result = [];
+        if (null === $filter) {
+            $result = array_keys($this->projections);
+            $result = array_slice($result, $offset, $limit);
+            sort($result, \SORT_NATURAL);
 
-        $skipped = 0;
-        $found = 0;
-
-        $projectionNames = array_keys($this->projections);
-        ksort($projectionNames);
+            return $result;
+        }
 
-        foreach ($projectionNames as $projectionName) {
-            if ($regex) {
-                if (! preg_match("/$filter/", $projectionName)) {
-                    continue;
-                }
+        if (isset($this->projections[$filter])) {
+            return [$filter];
+        }
 
-                $result[] = $projectionName;
-                ++$found;
-            } elseif (null === $filter || $filter === $projectionName) {
-                if ($offset > $skipped) {
-                    ++$skipped;
-                    continue;
-                }
+        return [];
+    }
 
-                $result[] = $projectionName;
-                ++$found;
-            }
+    public function fetchProjectionNamesRegex(string $regex, int $limit = 20, int $offset = 0): array
+    {
+        if (0 >= $limit) {
+            throw new Exception\OutOfRangeException(
+                'Invalid limit "'.$limit.'" given. Must be greater than 0.'
+            );
+        }
 
-            if ($found === $limit) {
-                break;
-            }
+        if (0 > $offset) {
+            throw new Exception\OutOfRangeException(
+                'Invalid offset "'.$offset.'" given. Must be greater or equal than 0.'
+            );
         }
 
-        return $result;
+        set_error_handler(function ($errorNo, $errorMsg): void {
+            throw new Exception\RuntimeException($errorMsg);
```

Comment:
Is the error message meaningful enough? With this: `preg_match('/invalid)/', '');` I receive the message: `Compilation failed: unmatched parentheses at offset 7` - Maybe it should be prepended with "Invalid regex: " ?